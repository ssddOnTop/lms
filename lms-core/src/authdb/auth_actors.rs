use anyhow::anyhow;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::str::FromStr;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum Authority {
    Admin,
    Faculty,
    Student,
}

impl FromStr for Authority {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        match s.as_str() {
            "admin" => Ok(Authority::Admin),
            "faculty" => Ok(Authority::Faculty),
            "student" => Ok(Authority::Student),
            _ => Err(anyhow!("Unable to serialize")),
        }
    }
}

impl Authority {
    pub fn from_int(int: u8) -> anyhow::Result<Self> {
        match int {
            0 => Ok(Authority::Admin),
            1 => Ok(Authority::Faculty),
            2 => Ok(Authority::Student),
            _ => Err(anyhow!("Unable to serialize")),
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct User {
    pub username: String,
    pub name: String,
    pub password: String,
    pub authority: Authority,
}

#[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq)]
pub struct Users {
    users: HashMap<String, User>,
}

impl Users {
    pub fn get(&self, username: &str) -> Option<User> {
        self.users.get(username).cloned()
    }
    pub fn insert(&mut self, user: User) {
        self.users.insert(user.username.clone(), user);
    }
}

#[cfg(test)]
mod tests {
    use crate::authdb::auth_actors::{Authority, User, Users};
    use std::collections::HashMap;

    #[test]
    fn test_ser() -> anyhow::Result<()> {
        let mut users = HashMap::new();
        users.insert(
            "foo".to_string(),
            User {
                username: "foo".to_string(),
                name: "Foo".to_string(),
                password: "foopassword".to_string(),
                authority: Authority::Admin,
            },
        );
        let users = Users { users };

        let ser = serde_json::to_string(&users)?;
        assert_eq!(
            ser,
            r#"{"users":{"foo":{"username":"foo","name":"Foo","password":"foopassword","authority":"Admin"}}}"#
        );
        Ok(())
    }
    #[test]
    fn test_deser() -> anyhow::Result<()> {
        let ser = r#"{"users":{"foo":{"username":"foo","name":"Foo","password":"foopassword","authority":"Admin"}}}"#;
        let mut users = HashMap::new();
        users.insert(
            "foo".to_string(),
            User {
                username: "foo".to_string(),
                name: "Foo".to_string(),
                password: "foopassword".to_string(),
                authority: Authority::Admin,
            },
        );
        let users = Users { users };
        assert_eq!(users, serde_json::from_str(ser)?);
        Ok(())
    }
}
