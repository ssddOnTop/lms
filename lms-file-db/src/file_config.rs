use serde::{Deserialize, Deserializer, Serialize, Serializer};
use anyhow::Result;
use base64::Engine;
use base64::prelude::BASE64_STANDARD;
use serde::ser::SerializeStruct;
use lms_core::authdb::auth_actors::Authority;
use lms_core::is_default;

#[derive(Clone)]
pub struct FileHolder {
    pub name: String,
    pub content: Vec<u8>,
}



pub struct InsertionInfo {
    pub title: String,
    pub description: String,
    pub timestamp: u128,
    pub end_time: Option<u128>,
    pub authority: Authority,
}

impl Serialize for FileHolder {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error> where S: Serializer {
        let mut state = serializer.serialize_struct("FileHolder", 2)?;
        state.serialize_field("name", &self.name)?;
        let b64 = BASE64_STANDARD.encode(&self.content);
        state.serialize_field("content", &b64)?;
        state.end()
    }
}

struct FileHolderVisitor;

impl<'de> serde::de::Visitor<'de> for FileHolderVisitor {
    type Value = FileHolder;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("struct FileHolder")
    }

    fn visit_map<A>(self, mut map: A) -> std::result::Result<Self::Value, A::Error> where A: serde::de::MapAccess<'de> {
        let mut name = None;
        let mut content = None;
        while let Some(key) = map.next_key()? {
            match key {
                "name" => {
                    if name.is_some() {
                        return Err(serde::de::Error::duplicate_field("name"));
                    }
                    name = Some(map.next_value()?);
                }
                "content" => {
                    if content.is_some() {
                        return Err(serde::de::Error::duplicate_field("content"));
                    }
                    let b64: String = map.next_value()?;
                    content = Some(BASE64_STANDARD.decode(b64.as_bytes()).map_err(serde::de::Error::custom)?);
                }
                _ => {
                    return Err(serde::de::Error::unknown_field(key, &["name", "content"]));
                }
            }
        }
        let name = name.ok_or_else(|| serde::de::Error::missing_field("name"))?;
        let content = content.ok_or_else(|| serde::de::Error::missing_field("content"))?;
        Ok(FileHolder { name, content })
    }
}

impl<'de> Deserialize<'de> for FileHolder {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error> where D: Deserializer<'de> {
        deserializer.deserialize_struct("FileHolder", &["name", "content"], FileHolderVisitor)
    }
}

#[derive(Serialize, Deserialize)]
pub struct RemoteFileConfig {
    #[serde(default, skip_serializing_if = "is_default")]
    pub files: Vec<FileHolder>, // file names and content
    pub metadata: Metadata,
}

#[derive(Serialize, Deserialize)]
pub struct LocalFileConfig {
    #[serde(default, skip_serializing_if = "is_default")]
    pub files: Vec<String>, // path of files in the same dir
    pub metadata: Metadata,
}

#[derive(Serialize, Deserialize)]
pub struct Metadata {
    pub title: String,
    pub description: String,
    pub timestamp: u128,
    #[serde(default, skip_serializing_if = "is_default")]
    pub end_time: Option<u128>,
    pub authority: Authority,
}

impl RemoteFileConfig {
    pub fn combine_info(insertion_info: InsertionInfo, files: Vec<FileHolder>) -> Self {
        Self {
            files,
            metadata: Metadata {
                title: insertion_info.title,
                description: insertion_info.description,
                timestamp: insertion_info.timestamp,
                end_time: insertion_info.end_time,
                authority: insertion_info.authority,
            },
        }
    }
    pub fn serialize(self) -> Result<String> {
        Ok(serde_json::to_string(&self)?)
    }
    pub fn deserialize(data: &str) -> Result<Self> {
        Ok(serde_json::from_str(data)?)
    }
}

impl LocalFileConfig {

    pub fn combine_info(insertion_info: InsertionInfo, files: &Vec<FileHolder>) -> Self {
        Self {
            files: files.iter().map(|file| file.name.clone()).collect(),
            metadata: Metadata {
                title: insertion_info.title,
                description: insertion_info.description,
                timestamp: insertion_info.timestamp,
                end_time: insertion_info.end_time,
                authority: insertion_info.authority,
            },
        }
    }

    pub fn serialize(self) -> Result<String> {
        Ok(serde_json::to_string(&self)?)
    }
    pub fn deserialize(data: &str) -> Result<Self> {
        Ok(serde_json::from_str(data)?)
    }
}