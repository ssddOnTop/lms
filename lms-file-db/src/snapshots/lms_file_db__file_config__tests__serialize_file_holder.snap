---
source: lms-file-db/src/file_config.rs
expression: serialized
---
{"name":"example.txt","content":"SGVsbG8sIHdvcmxkIQ=="}
